Source/WebCore/rendering/style/PositionArea.h

```diff
@@ -25,43 +25,76 @@
  */
 
 #pragma once
+#include "BoxSides.h"
+#include "WritingMode.h"
 
 namespace WebCore {
 
 // The axis that the span specifies.
+// Represented as 3 bits, with the last one indicating the corresponding [Logical]BoxAxis value.
 enum class PositionAreaAxis : uint8_t {
-    // Physical axes. Implies self is PositionAreaSelf::No, as physical axes
-    // do not depend on writing mode of any elements.
-    Horizontal,
-    Vertical,
-
-    // Logical axes.
-    X,
-    Y,
-    Block,
-    Inline,
-};
+    // Physical axes × Physical directions.
+    Horizontal = 0b000,
+    Vertical   = 0b001,
 
-// Specifies which tile(s) on the axis that the position-area span occupies.
-enum class PositionAreaTrack : uint8_t {
-    // First tile.
-    Start,
+    // Physical axes × Logical directions.
+    X = 0b010,
+    Y = 0b011,
 
-    // First and center tiles.
-    SpanStart,
+    // Logical axes × Logical directions.
+    Inline = 0b110,
+    Block  = 0b111,
+};
 
-    // Last tile.
-    End,
+static inline bool isPositionAreaAxisLogical(const PositionAreaAxis positionAxis)
+{
+    return static_cast<uint8_t>(positionAxis) & 0b100;
+}
+
+static inline bool isPositionAreaDirectionLogical(const PositionAreaAxis positionAxis)
+{
+    return static_cast<uint8_t>(positionAxis) & 0b010;
+}
+
+static inline BoxAxis mapPositionAreaAxisToPhysicalAxis(const PositionAreaAxis positionAxis, const WritingMode writingMode)
+{
+    auto physicalAxis = static_cast<uint8_t>(positionAxis) & 0b001;
+    if (isPositionAreaAxisLogical(positionAxis) && writingMode.isVertical())
+        physicalAxis = !physicalAxis;
+    return static_cast<BoxAxis>(physicalAxis);
+}
+
+static inline LogicalBoxAxis mapPositionAreaAxisToLogicalAxis(const PositionAreaAxis positionAxis, const WritingMode writingMode)
+{
+    auto logicalAxis = static_cast<uint8_t>(positionAxis) & 0b001;
+    if (!isPositionAreaAxisLogical(positionAxis) && writingMode.isVertical())
+        logicalAxis = !logicalAxis;
+    return static_cast<LogicalBoxAxis>(logicalAxis);
+}
+
+// Specifies which tracks(s) on the axis that the position-area span occupies.
+// Represented as 3 bits: start track, center track, end track.
+enum class PositionAreaTrack : uint8_t {
+    Start     = 0b001, // First track.
+    SpanStart = 0b011, // First and center tracks.
+    End       = 0b100, // Last track.
+    SpanEnd   = 0b110, // Center and last track.
+    Center    = 0b010, // Center track.
+    SpanAll   = 0b111, // All tracks along the axis.
+};
 
-    // Center and last tiles.
-    SpanEnd,
+static inline PositionAreaTrack invertPositionAreaTrack(PositionAreaTrack track)
```

Comment:
Very clever way to flip the tracks, although it took me a while to see the logic behind all the bitwises 😅 I'd certainly appreciate some more comments here.

Reply:
And I think "flip" is more appropriate than "invert"?

Reply:
Sounds good! Changed it over. Actually, we could even just call it `flip()`, since the type name isn't adding anything here.

Reply:
This is one of those moments where I wish C++ has enum methods so we could have a `PositionAreaTrack::flip` or something like that...

`flipPositionAreaTrack` is too verbose but I feel like `flip` is too ambiguous, so maybe `flipTrack`?

Reply:
I have many of those moments. :)

I think we should either go with `flip()` and rely on the type information, or go with `flipPositionAreaTrack()` and be explicit about it. It's not a commonly used function (there's probably only going to be somewhere between 1-3 calls in our codebase overall) so I don't think we need to worry about the length.

Reply:
+1 for `flipPositionAreaTrack`.